src/chat.rs

```diff
@@ -673,30 +673,30 @@ pub fn msgtype_has_file(msgtype: Viewtype) -> bool {
 
 fn prepare_msg_common(context: &Context, chat_id: u32, msg: &mut Message) -> Result<u32, Error> {
     msg.id = 0;
-
     if msg.type_0 == Viewtype::Text {
         // the caller should check if the message text is empty
     } else if msgtype_has_file(msg.type_0) {
-        let path_filename = msg.param.get(Param::File);
-
-        ensure!(
-            path_filename.is_some(),
-            "Attachment missing for message of type #{}.",
-            msg.type_0
-        );
-
-        let mut path_filename = path_filename.unwrap().to_string();
-
-        if msg.state == MessageState::OutPreparing && !dc_is_blobdir_path(context, &path_filename) {
-            bail!("Files must be created in the blob-directory.");
-        }
-
-        ensure!(
-            dc_make_rel_and_copy(context, &mut path_filename),
-            "Failed to copy"
-        );
-
-        msg.param.set(Param::File, &path_filename);
+        let blob = msg
+            .param
+            .get(Param::File)
+            .ok_or_else(|| format_err!("Attachment missing for message of type #{}.", msg.type_0))
+            .and_then(|param| ParamsFile::from_param(context, param).map_err(|e| e.into()))
+            .and_then(|file| match file {
```

Comment:
this would be nicer with using `?`

```rust
let param = msg.param./*...*/.ok_or_else()?;
let file = ParamsFile::from_param(context, param)?;
match file {
}
```

Reply:
depends on the mood i'm in... but sure.  It does make some of the error casting go away i guess.  May also alleviate some concerns @hpk42 had.